src/main/java/com/yahoo/bullet/typesystem/TypedObject.java

```diff
@@ -127,9 +135,149 @@ public int compareTo(TypedObject o) {
+     * @throws RuntimeException if not supported.
+     */
+    public Boolean containsValue(TypedObject target) {
+        switch (type) {
+            case LIST:
+                for (Object element : (List) value) {
+                    if (element instanceof Map) {
+                        if (containsValueInPrimitiveMap((Map) element, target)) {
+                            return true;
+                        }
+                    }
+                    // Support list of primitives after https://github.com/bullet-db/bullet-record/issues/12 is done.
+                }
+                return false;
+            case MAP:
+                for (Map.Entry entry : ((Map<?, ?>) value).entrySet()) {
+                    Object entryValue = entry.getValue();
+                    if (entryValue instanceof Map) {
+                        if (containsValueInPrimitiveMap((Map) entryValue, target)) {
+                            return true;
+                        }
+                    } else {
+                        TypedObject typedObject = new TypedObject(entryValue);
+                        if (typedObject.compareTo(target) == 0) {
+                            return true;
+                        }
+
+                    }
+                }
+                return false;
+            default:
+                throw new RuntimeException("Unsupported type cannot be operated CONTAINSVALUE: " + type);
+        }
+    }
+
     @Override
     public String toString() {
         return type == Type.NULL ? Type.NULL_EXPRESSION : value.toString();
     }
+
+    private static Boolean containsValueInPrimitiveMap(Map<?, ?> map, TypedObject target) {
+        for (Map.Entry entry : map.entrySet()) {
+            TypedObject typedObject = new TypedObject(entry.getValue());
+            if (typedObject.compareTo(target) == 0) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private static Type extractPrimitiveTypeFromMap(Map map) {
+        if (map.isEmpty()) {
+            return Type.UNKNOWN;
+        }
+        Object firstValue = map.get(map.keySet().iterator().next());
+        return Type.getType(firstValue);
+
+    }
+
+    private static Type extractPrimitiveType(Type type, Object target) {
+        switch (type) {
+            case LIST:
+                List list = (List) target;
+                if (list.isEmpty()) {
+                    return Type.UNKNOWN;
+                }
+                if (list.get(0) instanceof Map) {
+                    return extractPrimitiveTypeFromMap((Map) list.get(0));
+                }
+                return Type.getType(list.get(0));
+            case MAP:
+                Map map = (Map) target;
+                if (map.isEmpty()) {
+                    return Type.UNKNOWN;
+                }
+                Object firstValue = map.get(map.keySet().iterator().next());
```

Comment:
Object firstValue = map.values().iterator().next();

is a bit better, i think, unless it doesn't work for some reason.

Reply:
changed to map.values().stream().findAny().get()